### ISSUE TITLE
Investigate why the app takes so much memory on startup

### ISSUE THREAD

**Issue by jrainville** [status-im/status-desktop#10339]
## Description

The app takes a lot of memory and it is concerning.

Even on startup, with nothing touched, the app takes ~650 MB or RAM.

I do have the Status Community on that account, but it just opened one chat, so it shouldn't have loaded anything else except that channel, the list of communities, the channels from the current community, our settings, our contacts, probably some stuff for the wallet and the members of the community

Some guesses:
- Maybe some images are kept in memory
  - In theory it should use the localhost server to serve images from status-go
- Too much stuff loaded from the start?
  - Maybe the browser is loaded from the start or some wallet stuff even though we haven't opened it

## Steps to reproduce

Just open the app, preferably with the Status Community

#### Expected behavior

The memory should startspretty low on startup. Similar to Discord, which is ~250MB

#### Actual behavior

Memory starts at ~750MB on startup

### Additional Information

- Status desktop version: 0.11.0
- Operating System: Ubuntu

**Comment by caybro** [comment]
> Maybe some images are kept in memory

_ALL_ images are kept in memory due to the QML image cache that's `true` by default: https://doc.qt.io/qt-5/qml-qtquick-image.html#cache-prop

We have way too many resources (including all the icons, fonts, emojis and whatnot); all of them are in memory at the startup

**Comment by alexjba** [comment]
> @jrainville I now have around 480MB at startup, with some "strategic" `cache: false` in places where it's not desired:
> 
> ![image](https://user-images.githubusercontent.com/5377645/234536139-17f90b2c-8580-4d54-9f4d-57c664f128fe.png)
> 
> Worth submitting a PR for this?

I'm all for it if improves the performance, but some highly used resources like icons could remain cached. They're small in size and widely used in the app.

How much RAM can be saved? 
I've done some tests a while ago with caching disabled on all images, but I couldn't see a big improvement in RAM usage. A few MB at most. But that was on QT 5.14 if I remember correctly.

One thing we can consider is to never cache the images received from status-go or http requests. Their size can skyrocket the RAM usage. Mostly the GIFs. Unfortunately we display GIFs using sockets and qml cannot play the img sequence with caching disabled, so a little refactoring might be required if it improves the RAM usage.

> The memory should startspretty low on startup. Similar to Discord, which is ~250MB

@jrainville Lucky we're not that far from it IMO 😄. Discord is multi-process and the total RAM usage I see is around 550MB (post-login). VS 790 MB used by Status.

**Comment by caybro** [comment]
> I'm all for it if improves the performance, but some highly used resources like icons could remain cached. They're small in size and widely used in the app.

Yeah, those are still cached

> One thing we can consider is to never cache the images received from status-go or http requests. Their size can skyrocket the RAM usage. Mostly the GIFs. Unfortunately we display GIFs using sockets and qml cannot play the img sequence with caching disabled, so a little refactoring might be required if it improves the RAM usage.

Exactly what I did locally

**Comment by jrainville** [comment]
> Worth submitting a PR for this?

@caybro yeah for sure. I'm curious to see what you did

**Comment by caybro** [comment]
> > Worth submitting a PR for this?
> 
> @caybro yeah for sure. I'm curious to see what you did

https://github.com/status-im/status-desktop/pull/10444

**Comment by iurimatias** [comment]
not ready for 0.12, moving to 0.13

**Comment by osmaczko** [comment]
Good starting point would be to investigate all the responses we get from status-go to see if we get plain assets data instead of local media-server urls.

**Comment by jrainville** [comment]
> Good starting point would be to investigate all the responses we get from status-go to see if we get plain assets data instead of local media-server urls.

Yup, that's what I meant when I wrote: "Maybe some images are kept in memory
In theory it should use the localhost server to serve images from status-go"

**Comment by alexjba** [comment]
> > Good starting point would be to investigate all the responses we get from status-go to see if we get plain assets data instead of local media-server urls.
> 
> I'm betting all we get is the binary blobs, ie. `data:image/jpeg;base64,...`

Seems like a valid point. I've installed an event filter to sniff into the `QQuickImageBase` objects and there's quite a few using base64 data as source and all of them are cached.
I don't have the flows yet, but looks like token images and profile images.

**Comment by alexjba** [comment]
#### I'm hitting a dead end here!

After login I have between 700-900MB usage
 - status-go pprof reports about 60MB heap size
 - nim reports about 30 MB heap, 90MB total memory
 - QML seems to account for about ~150MB after login. Tested by observing the difference after removing the `appLayout` items from `appMain.qml`.
 
 **There are at least 300MB unaccounted for. The `heap` tool reports about the same memory size dedicated to `non-object`**

**See the docs attached below**
 
 **Tools used to investigate:**
 1. `Instruments` - Reports lots of memory leaks in Qt (we were almost leak free after the previous perf improvements). But the total size is small. A few kb
 2. `heap` tool. As expected, Qt generates the most memory usage. But there are about 300MB dedicated to `non-objects`
 3. `pprof` to profile status-go. Didn't notice anything worth invetigating
 4.  nim `dumpNumberOfInstances`, `GC_getStatistics`. The reported memory usage is much smaller than I've expected. Not sure I can trust these numbers because they represent only what the GC is tracking.
 
 ### Next steps

**Nim**
- In-depth analysis of nim memory usage to see if:
   1 - the reported `non-objects` are related to nim runtime. Not sure how to analyse this memory. I hope Valgrind is more helpful, but it needs to be analysed on Linux.
   2- if the JsonNode could be one of the memory offenders. JsonNode has a high memory footprint. In my tests the memory used ranged from 3x to 5x the string representation of the json. There is a project benchmarking the JsonNode (https://github.com/Araq/packedjson) and proposing an alternative implementation. From my analysis the JsonNode objects are transitory, used just a mean to parse the string.
@status-im/desktop I'd really appreciate some thoughts/help here. Can anyone on Linux run Valgrind to see if there is something useful in the reports?

#### Memory usage increasing and not being released over time

**status-go**
Use the local server to serve the following images. There's lots of base64 images being transferred from status-go and kept in memory in nim and qt:
 - Community banner
- User profile images (E.g. my profile, demo user profile in Appearance settings)
- Community tokens - The community settings view is always loaded and all the community tokens as well. TODO: use lazy load.
- Community logo
- Profile QR codes
- wallet community tokens. The `Collectibles` view is already a memory bomb because we're missing the means to limit the external collectible image size (we could easily load hundreds of mb for 1 fancy collectible).

[nim-heap.txt](https://github.com/status-im/status-desktop/files/13726040/nim-heap.txt) - `dumpNumberOfInstances` results
[nim-stats.txt](https://github.com/status-im/status-desktop/files/13726041/nim-stats.txt) - `GC_getStatistics `
[heap.txt](https://github.com/status-im/status-desktop/files/13726060/heap.txt) - global heap result
Pprof results:
<img width="1630" alt="Screenshot 2023-12-19 at 14 36 47" src="https://github.com/status-im/status-desktop/assets/47811206/a6a68ecd-0960-45b8-961e-b610348e5a12">

**Comment by alexjba** [comment]
> FYI @dlipicar just yesterday moved collectibles images to media server, that should decrease a lot. I also have PR with community token images to use media server instead.

Thanks! Looking forward to see the improvements!
I've created this [task](https://github.com/status-im/status-desktop/issues/13061) to serve the images using the media server. But now I think it's more of an Epic in this case, to track the progress. Could you please help by linking the current tasks for the collectibles images to this one?

> From my experience dealing with memory in Qt. Qt application itself allocates around 50-60MB. Each first usage of specific module loads new library. For ex. first usage of Image in QML will bump memory usage by ~15MB because whole library is loaded and it stays forever in memory. Additionally QML is slow with releasing memory for textures, as they are optimized to be destroyed in separate thread.

Good point! I can also look for unnecessary dependencies, thanks!

Indeed, Qt can use a lot of memory and it really depends on a lot of factors (E.g. increasing the window size could easily eat-up 50MB of memory). But that's known and somehow expected. While breaking down the memory, there's still lots of memory unaccounted (almost half the memory used after login) for and I can't find who's allocating it. Could be that other tools like Valgrind to offer a better picture.
I assume there's also another reason for the high memory usage and I'm suspecting the nim layer because 1- it acts like a never-ending cache 2 - has the potential to eat RAM like peanuts (see the JsonNode implementation as an example).

**Comment by dlipicar** [comment]
> FYI @dlipicar just yesterday moved collectibles images to media server, that should decrease a lot. I also have PR with community token images to use media server instead.

Still, that's only for community collectibles. There could still be a normal collectible linking to a 1GB video in its metadata and we would try to download and play that... I've discussed this with Ben to see if we can do something like play animations on hover only or something like that to reduce both memory and CPU usage.

**Comment by alexandraB99** [comment]
closing - follow up task https://github.com/status-im/status-desktop/issues/13061.
